// @deck.gl/widgets/dist/dist.dev.js
(function webpackUniversalModuleDefinition(root, factory) {
  if (typeof exports === 'object' && typeof module === 'object')
    module.exports = factory();
  else if (typeof define === 'function' && define.amd) define([], factory);
        else if (typeof exports === 'object') exports['deck'] = factory();
  else root['deck'] = factory();})(globalThis, function () {
"use strict";
var __exports__ = (() => {
  var __create = Object.create;
  var __defProp = Object.defineProperty;
  var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __getProtoOf = Object.getPrototypeOf;
  var __hasOwnProp = Object.prototype.hasOwnProperty;
  var __commonJS = (cb, mod) => function __require() {
    return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
  };
  var __export = (target, all) => {
    for (var name in all)
      __defProp(target, name, { get: all[name], enumerable: true });
  };
  var __copyProps = (to, from, except, desc) => {
    if (from && typeof from === "object" || typeof from === "function") {
      for (let key of __getOwnPropNames(from))
        if (!__hasOwnProp.call(to, key) && key !== except)
          __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
    }
    return to;
  };
  var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
  var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
    // If the importer is in node compatibility mode or this is not an ESM
    // file that has been converted to a CommonJS file using a Babel-
    // compatible transform (i.e. "__esModule" has not been set), then set
    // "default" to the CommonJS "module.exports" for node compatibility.
    isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
    mod
  ));
  var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

  // external-global-plugin:@deck.gl/core
  var require_core = __commonJS({
    "external-global-plugin:@deck.gl/core"(exports, module) {
      module.exports = globalThis.deck;
    }
  });

  // bundle.ts
  var bundle_exports = {};
  __export(bundle_exports, {
    CompassWidget: () => CompassWidget,
    DarkGlassTheme: () => DarkGlassTheme,
    DarkTheme: () => DarkTheme,
    FullscreenWidget: () => FullscreenWidget,
    LightGlassTheme: () => LightGlassTheme,
    LightTheme: () => LightTheme,
    ResetViewWidget: () => ResetViewWidget,
    ScreenshotWidget: () => ScreenshotWidget,
    ZoomWidget: () => ZoomWidget,
    _WidgetImpl: () => WidgetImpl,
    _components: () => components_exports
  });

  // ../core/bundle/peer-dependency.ts
  var peer_dependency_exports = {};
  var import_core = __toESM(require_core(), 1);
  __reExport(peer_dependency_exports, __toESM(require_core(), 1));
  if (!import_core.Layer) {
    throw new Error("@deck.gl/core is not found");
  }

  // bundle.ts
  __reExport(bundle_exports, peer_dependency_exports);

  // src/fullscreen-widget.tsx
  var import_core2 = __toESM(require_core(), 1);

  // ../../node_modules/preact/dist/preact.module.js
  var n;
  var l;
  var u;
  var t;
  var i;
  var o;
  var r;
  var f;
  var e;
  var c;
  var s;
  var a;
  var h = {};
  var p = [];
  var v = /acit|ex(?:s|g|n|p|$)|rph|grid|ows|mnc|ntw|ine[ch]|zoo|^ord|itera/i;
  var y = Array.isArray;
  function d(n2, l2) {
    for (var u3 in l2)
      n2[u3] = l2[u3];
    return n2;
  }
  function w(n2) {
    var l2 = n2.parentNode;
    l2 && l2.removeChild(n2);
  }
  function _(l2, u3, t2) {
    var i3, o2, r2, f3 = {};
    for (r2 in u3)
      "key" == r2 ? i3 = u3[r2] : "ref" == r2 ? o2 = u3[r2] : f3[r2] = u3[r2];
    if (arguments.length > 2 && (f3.children = arguments.length > 3 ? n.call(arguments, 2) : t2), "function" == typeof l2 && null != l2.defaultProps)
      for (r2 in l2.defaultProps)
        void 0 === f3[r2] && (f3[r2] = l2.defaultProps[r2]);
    return g(l2, f3, i3, o2, null);
  }
  function g(n2, t2, i3, o2, r2) {
    var f3 = { type: n2, props: t2, key: i3, ref: o2, __k: null, __: null, __b: 0, __e: null, __d: void 0, __c: null, constructor: void 0, __v: null == r2 ? ++u : r2, __i: -1, __u: 0 };
    return null == r2 && null != l.vnode && l.vnode(f3), f3;
  }
  function k(n2) {
    return n2.children;
  }
  function b(n2, l2) {
    this.props = n2, this.context = l2;
  }
  function x(n2, l2) {
    if (null == l2)
      return n2.__ ? x(n2.__, n2.__i + 1) : null;
    for (var u3; l2 < n2.__k.length; l2++)
      if (null != (u3 = n2.__k[l2]) && null != u3.__e)
        return u3.__e;
    return "function" == typeof n2.type ? x(n2) : null;
  }
  function C(n2) {
    var l2, u3;
    if (null != (n2 = n2.__) && null != n2.__c) {
      for (n2.__e = n2.__c.base = null, l2 = 0; l2 < n2.__k.length; l2++)
        if (null != (u3 = n2.__k[l2]) && null != u3.__e) {
          n2.__e = n2.__c.base = u3.__e;
          break;
        }
      return C(n2);
    }
  }
  function M(n2) {
    (!n2.__d && (n2.__d = true) && i.push(n2) && !P.__r++ || o !== l.debounceRendering) && ((o = l.debounceRendering) || r)(P);
  }
  function P() {
    var n2, u3, t2, o2, r2, e2, c2, s2;
    for (i.sort(f); n2 = i.shift(); )
      n2.__d && (u3 = i.length, o2 = void 0, e2 = (r2 = (t2 = n2).__v).__e, c2 = [], s2 = [], t2.__P && ((o2 = d({}, r2)).__v = r2.__v + 1, l.vnode && l.vnode(o2), O(t2.__P, o2, r2, t2.__n, t2.__P.namespaceURI, 32 & r2.__u ? [e2] : null, c2, null == e2 ? x(r2) : e2, !!(32 & r2.__u), s2), o2.__v = r2.__v, o2.__.__k[o2.__i] = o2, j(c2, o2, s2), o2.__e != e2 && C(o2)), i.length > u3 && i.sort(f));
    P.__r = 0;
  }
  function S(n2, l2, u3, t2, i3, o2, r2, f3, e2, c2, s2) {
    var a2, v2, y2, d2, w2, _2 = t2 && t2.__k || p, g2 = l2.length;
    for (u3.__d = e2, $(u3, l2, _2), e2 = u3.__d, a2 = 0; a2 < g2; a2++)
      null != (y2 = u3.__k[a2]) && "boolean" != typeof y2 && "function" != typeof y2 && (v2 = -1 === y2.__i ? h : _2[y2.__i] || h, y2.__i = a2, O(n2, y2, v2, i3, o2, r2, f3, e2, c2, s2), d2 = y2.__e, y2.ref && v2.ref != y2.ref && (v2.ref && N(v2.ref, null, y2), s2.push(y2.ref, y2.__c || d2, y2)), null == w2 && null != d2 && (w2 = d2), 65536 & y2.__u || v2.__k === y2.__k ? e2 = I(y2, e2, n2) : "function" == typeof y2.type && void 0 !== y2.__d ? e2 = y2.__d : d2 && (e2 = d2.nextSibling), y2.__d = void 0, y2.__u &= -196609);
    u3.__d = e2, u3.__e = w2;
  }
  function $(n2, l2, u3) {
    var t2, i3, o2, r2, f3, e2 = l2.length, c2 = u3.length, s2 = c2, a2 = 0;
    for (n2.__k = [], t2 = 0; t2 < e2; t2++)
      r2 = t2 + a2, null != (i3 = n2.__k[t2] = null == (i3 = l2[t2]) || "boolean" == typeof i3 || "function" == typeof i3 ? null : "string" == typeof i3 || "number" == typeof i3 || "bigint" == typeof i3 || i3.constructor == String ? g(null, i3, null, null, null) : y(i3) ? g(k, { children: i3 }, null, null, null) : void 0 === i3.constructor && i3.__b > 0 ? g(i3.type, i3.props, i3.key, i3.ref ? i3.ref : null, i3.__v) : i3) ? (i3.__ = n2, i3.__b = n2.__b + 1, f3 = L(i3, u3, r2, s2), i3.__i = f3, o2 = null, -1 !== f3 && (s2--, (o2 = u3[f3]) && (o2.__u |= 131072)), null == o2 || null === o2.__v ? (-1 == f3 && a2--, "function" != typeof i3.type && (i3.__u |= 65536)) : f3 !== r2 && (f3 == r2 - 1 ? a2-- : f3 == r2 + 1 ? a2++ : f3 > r2 ? s2 > e2 - r2 ? a2 += f3 - r2 : a2-- : f3 < r2 && (f3 == r2 - a2 ? a2 -= f3 - r2 : a2++), f3 !== t2 + a2 && (i3.__u |= 65536))) : (o2 = u3[r2]) && null == o2.key && o2.__e && 0 == (131072 & o2.__u) && (o2.__e == n2.__d && (n2.__d = x(o2)), V(o2, o2, false), u3[r2] = null, s2--);
    if (s2)
      for (t2 = 0; t2 < c2; t2++)
        null != (o2 = u3[t2]) && 0 == (131072 & o2.__u) && (o2.__e == n2.__d && (n2.__d = x(o2)), V(o2, o2));
  }
  function I(n2, l2, u3) {
    var t2, i3;
    if ("function" == typeof n2.type) {
      for (t2 = n2.__k, i3 = 0; t2 && i3 < t2.length; i3++)
        t2[i3] && (t2[i3].__ = n2, l2 = I(t2[i3], l2, u3));
      return l2;
    }
    n2.__e != l2 && (l2 && n2.type && !u3.contains(l2) && (l2 = x(n2)), u3.insertBefore(n2.__e, l2 || null), l2 = n2.__e);
    do {
      l2 = l2 && l2.nextSibling;
    } while (null != l2 && 8 === l2.nodeType);
    return l2;
  }
  function L(n2, l2, u3, t2) {
    var i3 = n2.key, o2 = n2.type, r2 = u3 - 1, f3 = u3 + 1, e2 = l2[u3];
    if (null === e2 || e2 && i3 == e2.key && o2 === e2.type && 0 == (131072 & e2.__u))
      return u3;
    if (t2 > (null != e2 && 0 == (131072 & e2.__u) ? 1 : 0))
      for (; r2 >= 0 || f3 < l2.length; ) {
        if (r2 >= 0) {
          if ((e2 = l2[r2]) && 0 == (131072 & e2.__u) && i3 == e2.key && o2 === e2.type)
            return r2;
          r2--;
        }
        if (f3 < l2.length) {
          if ((e2 = l2[f3]) && 0 == (131072 & e2.__u) && i3 == e2.key && o2 === e2.type)
            return f3;
          f3++;
        }
      }
    return -1;
  }
  function T(n2, l2, u3) {
    "-" === l2[0] ? n2.setProperty(l2, null == u3 ? "" : u3) : n2[l2] = null == u3 ? "" : "number" != typeof u3 || v.test(l2) ? u3 : u3 + "px";
  }
  function A(n2, l2, u3, t2, i3) {
    var o2;
    n:
      if ("style" === l2)
        if ("string" == typeof u3)
          n2.style.cssText = u3;
        else {
          if ("string" == typeof t2 && (n2.style.cssText = t2 = ""), t2)
            for (l2 in t2)
              u3 && l2 in u3 || T(n2.style, l2, "");
          if (u3)
            for (l2 in u3)
              t2 && u3[l2] === t2[l2] || T(n2.style, l2, u3[l2]);
        }
      else if ("o" === l2[0] && "n" === l2[1])
        o2 = l2 !== (l2 = l2.replace(/(PointerCapture)$|Capture$/i, "$1")), l2 = l2.toLowerCase() in n2 || "onFocusOut" === l2 || "onFocusIn" === l2 ? l2.toLowerCase().slice(2) : l2.slice(2), n2.l || (n2.l = {}), n2.l[l2 + o2] = u3, u3 ? t2 ? u3.u = t2.u : (u3.u = e, n2.addEventListener(l2, o2 ? s : c, o2)) : n2.removeEventListener(l2, o2 ? s : c, o2);
      else {
        if ("http://www.w3.org/2000/svg" == i3)
          l2 = l2.replace(/xlink(H|:h)/, "h").replace(/sName$/, "s");
        else if ("width" != l2 && "height" != l2 && "href" != l2 && "list" != l2 && "form" != l2 && "tabIndex" != l2 && "download" != l2 && "rowSpan" != l2 && "colSpan" != l2 && "role" != l2 && "popover" != l2 && l2 in n2)
          try {
            n2[l2] = null == u3 ? "" : u3;
            break n;
          } catch (n3) {
          }
        "function" == typeof u3 || (null == u3 || false === u3 && "-" !== l2[4] ? n2.removeAttribute(l2) : n2.setAttribute(l2, "popover" == l2 && 1 == u3 ? "" : u3));
      }
  }
  function F(n2) {
    return function(u3) {
      if (this.l) {
        var t2 = this.l[u3.type + n2];
        if (null == u3.t)
          u3.t = e++;
        else if (u3.t < t2.u)
          return;
        return t2(l.event ? l.event(u3) : u3);
      }
    };
  }
  function O(n2, u3, t2, i3, o2, r2, f3, e2, c2, s2) {
    var a2, h3, p2, v2, w2, _2, g2, m, x2, C2, M2, P2, $2, I2, H, L2, T2 = u3.type;
    if (void 0 !== u3.constructor)
      return null;
    128 & t2.__u && (c2 = !!(32 & t2.__u), r2 = [e2 = u3.__e = t2.__e]), (a2 = l.__b) && a2(u3);
    n:
      if ("function" == typeof T2)
        try {
          if (m = u3.props, x2 = "prototype" in T2 && T2.prototype.render, C2 = (a2 = T2.contextType) && i3[a2.__c], M2 = a2 ? C2 ? C2.props.value : a2.__ : i3, t2.__c ? g2 = (h3 = u3.__c = t2.__c).__ = h3.__E : (x2 ? u3.__c = h3 = new T2(m, M2) : (u3.__c = h3 = new b(m, M2), h3.constructor = T2, h3.render = q), C2 && C2.sub(h3), h3.props = m, h3.state || (h3.state = {}), h3.context = M2, h3.__n = i3, p2 = h3.__d = true, h3.__h = [], h3._sb = []), x2 && null == h3.__s && (h3.__s = h3.state), x2 && null != T2.getDerivedStateFromProps && (h3.__s == h3.state && (h3.__s = d({}, h3.__s)), d(h3.__s, T2.getDerivedStateFromProps(m, h3.__s))), v2 = h3.props, w2 = h3.state, h3.__v = u3, p2)
            x2 && null == T2.getDerivedStateFromProps && null != h3.componentWillMount && h3.componentWillMount(), x2 && null != h3.componentDidMount && h3.__h.push(h3.componentDidMount);
          else {
            if (x2 && null == T2.getDerivedStateFromProps && m !== v2 && null != h3.componentWillReceiveProps && h3.componentWillReceiveProps(m, M2), !h3.__e && (null != h3.shouldComponentUpdate && false === h3.shouldComponentUpdate(m, h3.__s, M2) || u3.__v === t2.__v)) {
              for (u3.__v !== t2.__v && (h3.props = m, h3.state = h3.__s, h3.__d = false), u3.__e = t2.__e, u3.__k = t2.__k, u3.__k.forEach(function(n3) {
                n3 && (n3.__ = u3);
              }), P2 = 0; P2 < h3._sb.length; P2++)
                h3.__h.push(h3._sb[P2]);
              h3._sb = [], h3.__h.length && f3.push(h3);
              break n;
            }
            null != h3.componentWillUpdate && h3.componentWillUpdate(m, h3.__s, M2), x2 && null != h3.componentDidUpdate && h3.__h.push(function() {
              h3.componentDidUpdate(v2, w2, _2);
            });
          }
          if (h3.context = M2, h3.props = m, h3.__P = n2, h3.__e = false, $2 = l.__r, I2 = 0, x2) {
            for (h3.state = h3.__s, h3.__d = false, $2 && $2(u3), a2 = h3.render(h3.props, h3.state, h3.context), H = 0; H < h3._sb.length; H++)
              h3.__h.push(h3._sb[H]);
            h3._sb = [];
          } else
            do {
              h3.__d = false, $2 && $2(u3), a2 = h3.render(h3.props, h3.state, h3.context), h3.state = h3.__s;
            } while (h3.__d && ++I2 < 25);
          h3.state = h3.__s, null != h3.getChildContext && (i3 = d(d({}, i3), h3.getChildContext())), x2 && !p2 && null != h3.getSnapshotBeforeUpdate && (_2 = h3.getSnapshotBeforeUpdate(v2, w2)), S(n2, y(L2 = null != a2 && a2.type === k && null == a2.key ? a2.props.children : a2) ? L2 : [L2], u3, t2, i3, o2, r2, f3, e2, c2, s2), h3.base = u3.__e, u3.__u &= -161, h3.__h.length && f3.push(h3), g2 && (h3.__E = h3.__ = null);
        } catch (n3) {
          if (u3.__v = null, c2 || null != r2) {
            for (u3.__u |= c2 ? 160 : 32; e2 && 8 === e2.nodeType && e2.nextSibling; )
              e2 = e2.nextSibling;
            r2[r2.indexOf(e2)] = null, u3.__e = e2;
          } else
            u3.__e = t2.__e, u3.__k = t2.__k;
          l.__e(n3, u3, t2);
        }
      else
        null == r2 && u3.__v === t2.__v ? (u3.__k = t2.__k, u3.__e = t2.__e) : u3.__e = z(t2.__e, u3, t2, i3, o2, r2, f3, c2, s2);
    (a2 = l.diffed) && a2(u3);
  }
  function j(n2, u3, t2) {
    u3.__d = void 0;
    for (var i3 = 0; i3 < t2.length; i3++)
      N(t2[i3], t2[++i3], t2[++i3]);
    l.__c && l.__c(u3, n2), n2.some(function(u4) {
      try {
        n2 = u4.__h, u4.__h = [], n2.some(function(n3) {
          n3.call(u4);
        });
      } catch (n3) {
        l.__e(n3, u4.__v);
      }
    });
  }
  function z(l2, u3, t2, i3, o2, r2, f3, e2, c2) {
    var s2, a2, p2, v2, d2, _2, g2, m = t2.props, k2 = u3.props, b2 = u3.type;
    if ("svg" === b2 ? o2 = "http://www.w3.org/2000/svg" : "math" === b2 ? o2 = "http://www.w3.org/1998/Math/MathML" : o2 || (o2 = "http://www.w3.org/1999/xhtml"), null != r2) {
      for (s2 = 0; s2 < r2.length; s2++)
        if ((d2 = r2[s2]) && "setAttribute" in d2 == !!b2 && (b2 ? d2.localName === b2 : 3 === d2.nodeType)) {
          l2 = d2, r2[s2] = null;
          break;
        }
    }
    if (null == l2) {
      if (null === b2)
        return document.createTextNode(k2);
      l2 = document.createElementNS(o2, b2, k2.is && k2), r2 = null, e2 = false;
    }
    if (null === b2)
      m === k2 || e2 && l2.data === k2 || (l2.data = k2);
    else {
      if (r2 = r2 && n.call(l2.childNodes), m = t2.props || h, !e2 && null != r2)
        for (m = {}, s2 = 0; s2 < l2.attributes.length; s2++)
          m[(d2 = l2.attributes[s2]).name] = d2.value;
      for (s2 in m)
        if (d2 = m[s2], "children" == s2)
          ;
        else if ("dangerouslySetInnerHTML" == s2)
          p2 = d2;
        else if ("key" !== s2 && !(s2 in k2)) {
          if ("value" == s2 && "defaultValue" in k2 || "checked" == s2 && "defaultChecked" in k2)
            continue;
          A(l2, s2, null, d2, o2);
        }
      for (s2 in k2)
        d2 = k2[s2], "children" == s2 ? v2 = d2 : "dangerouslySetInnerHTML" == s2 ? a2 = d2 : "value" == s2 ? _2 = d2 : "checked" == s2 ? g2 = d2 : "key" === s2 || e2 && "function" != typeof d2 || m[s2] === d2 || A(l2, s2, d2, m[s2], o2);
      if (a2)
        e2 || p2 && (a2.__html === p2.__html || a2.__html === l2.innerHTML) || (l2.innerHTML = a2.__html), u3.__k = [];
      else if (p2 && (l2.innerHTML = ""), S(l2, y(v2) ? v2 : [v2], u3, t2, i3, "foreignObject" === b2 ? "http://www.w3.org/1999/xhtml" : o2, r2, f3, r2 ? r2[0] : t2.__k && x(t2, 0), e2, c2), null != r2)
        for (s2 = r2.length; s2--; )
          null != r2[s2] && w(r2[s2]);
      e2 || (s2 = "value", void 0 !== _2 && (_2 !== l2[s2] || "progress" === b2 && !_2 || "option" === b2 && _2 !== m[s2]) && A(l2, s2, _2, m[s2], o2), s2 = "checked", void 0 !== g2 && g2 !== l2[s2] && A(l2, s2, g2, m[s2], o2));
    }
    return l2;
  }
  function N(n2, u3, t2) {
    try {
      if ("function" == typeof n2) {
        var i3 = "function" == typeof n2.__u;
        i3 && n2.__u(), i3 && null == u3 || (n2.__u = n2(u3));
      } else
        n2.current = u3;
    } catch (n3) {
      l.__e(n3, t2);
    }
  }
  function V(n2, u3, t2) {
    var i3, o2;
    if (l.unmount && l.unmount(n2), (i3 = n2.ref) && (i3.current && i3.current !== n2.__e || N(i3, null, u3)), null != (i3 = n2.__c)) {
      if (i3.componentWillUnmount)
        try {
          i3.componentWillUnmount();
        } catch (n3) {
          l.__e(n3, u3);
        }
      i3.base = i3.__P = null;
    }
    if (i3 = n2.__k)
      for (o2 = 0; o2 < i3.length; o2++)
        i3[o2] && V(i3[o2], u3, t2 || "function" != typeof n2.type);
    t2 || null == n2.__e || w(n2.__e), n2.__c = n2.__ = n2.__e = n2.__d = void 0;
  }
  function q(n2, l2, u3) {
    return this.constructor(n2, u3);
  }
  function B(u3, t2, i3) {
    var o2, r2, f3, e2;
    l.__ && l.__(u3, t2), r2 = (o2 = "function" == typeof i3) ? null : i3 && i3.__k || t2.__k, f3 = [], e2 = [], O(t2, u3 = (!o2 && i3 || t2).__k = _(k, null, [u3]), r2 || h, h, t2.namespaceURI, !o2 && i3 ? [i3] : r2 ? null : t2.firstChild ? n.call(t2.childNodes) : null, f3, !o2 && i3 ? i3 : r2 ? r2.__e : t2.firstChild, o2, e2), j(f3, u3, e2);
  }
  n = p.slice, l = { __e: function(n2, l2, u3, t2) {
    for (var i3, o2, r2; l2 = l2.__; )
      if ((i3 = l2.__c) && !i3.__)
        try {
          if ((o2 = i3.constructor) && null != o2.getDerivedStateFromError && (i3.setState(o2.getDerivedStateFromError(n2)), r2 = i3.__d), null != i3.componentDidCatch && (i3.componentDidCatch(n2, t2 || {}), r2 = i3.__d), r2)
            return i3.__E = i3;
        } catch (l3) {
          n2 = l3;
        }
    throw n2;
  } }, u = 0, t = function(n2) {
    return null != n2 && null == n2.constructor;
  }, b.prototype.setState = function(n2, l2) {
    var u3;
    u3 = null != this.__s && this.__s !== this.state ? this.__s : this.__s = d({}, this.state), "function" == typeof n2 && (n2 = n2(d({}, u3), this.props)), n2 && d(u3, n2), null != n2 && this.__v && (l2 && this._sb.push(l2), M(this));
  }, b.prototype.forceUpdate = function(n2) {
    this.__v && (this.__e = true, n2 && this.__h.push(n2), M(this));
  }, b.prototype.render = k, i = [], r = "function" == typeof Promise ? Promise.prototype.then.bind(Promise.resolve()) : setTimeout, f = function(n2, l2) {
    return n2.__v.__b - l2.__v.__b;
  }, P.__r = 0, e = 0, c = F(false), s = F(true), a = 0;

  // src/components.tsx
  var components_exports = {};
  __export(components_exports, {
    ButtonGroup: () => ButtonGroup,
    GroupedIconButton: () => GroupedIconButton,
    IconButton: () => IconButton
  });

  // ../../node_modules/preact/jsx-runtime/dist/jsxRuntime.module.js
  var f2 = 0;
  var i2 = Array.isArray;
  function u2(e2, t2, n2, o2, i3, u3) {
    t2 || (t2 = {});
    var a2, c2, p2 = t2;
    if ("ref" in p2)
      for (c2 in p2 = {}, t2)
        "ref" == c2 ? a2 = t2[c2] : p2[c2] = t2[c2];
    var l2 = { type: e2, props: p2, key: n2, ref: a2, __k: null, __: null, __b: 0, __e: null, __d: void 0, __c: null, constructor: void 0, __v: --f2, __i: -1, __u: 0, __source: i3, __self: u3 };
    if ("function" == typeof e2 && (a2 = e2.defaultProps))
      for (c2 in a2)
        void 0 === p2[c2] && (p2[c2] = a2[c2]);
    return l.vnode && l.vnode(l2), l2;
  }

  // src/components.tsx
  var IconButton = (props) => {
    const { className, label, onClick } = props;
    return /* @__PURE__ */ u2("div", { className: "deck-widget-button", children: /* @__PURE__ */ u2(
      "button",
      {
        className: `deck-widget-icon-button ${className}`,
        type: "button",
        onClick,
        title: label,
        children: /* @__PURE__ */ u2("div", { className: "deck-widget-icon" })
      }
    ) });
  };
  var ButtonGroup = (props) => {
    const { children, orientation } = props;
    return /* @__PURE__ */ u2("div", { className: `deck-widget-button-group ${orientation}`, children });
  };
  var GroupedIconButton = (props) => {
    const { className, label, onClick } = props;
    return /* @__PURE__ */ u2(
      "button",
      {
        className: `deck-widget-icon-button ${className}`,
        type: "button",
        onClick,
        title: label,
        children: /* @__PURE__ */ u2("div", { className: "deck-widget-icon" })
      }
    );
  };

  // src/fullscreen-widget.tsx
  var FullscreenWidget = class {
    constructor(props) {
      this.id = "fullscreen";
      this.placement = "top-left";
      this.fullscreen = false;
      this.id = props.id ?? this.id;
      this.placement = props.placement ?? this.placement;
      this.props = {
        ...props,
        enterLabel: props.enterLabel ?? "Enter Fullscreen",
        exitLabel: props.exitLabel ?? "Exit Fullscreen",
        style: props.style ?? {}
      };
    }
    onAdd({ deck }) {
      const { style, className } = this.props;
      const el = document.createElement("div");
      el.classList.add("deck-widget", "deck-widget-fullscreen");
      if (className)
        el.classList.add(className);
      (0, import_core2._applyStyles)(el, style);
      this.deck = deck;
      this.element = el;
      this.update();
      document.addEventListener("fullscreenchange", this.onFullscreenChange.bind(this));
      return el;
    }
    onRemove() {
      this.deck = void 0;
      this.element = void 0;
      document.removeEventListener("fullscreenchange", this.onFullscreenChange.bind(this));
    }
    update() {
      const { enterLabel, exitLabel } = this.props;
      const element = this.element;
      if (!element) {
        return;
      }
      const ui = /* @__PURE__ */ u2(
        IconButton,
        {
          onClick: this.handleClick.bind(this),
          label: this.fullscreen ? exitLabel : enterLabel,
          className: this.fullscreen ? "deck-widget-fullscreen-exit" : "deck-widget-fullscreen-enter"
        }
      );
      B(ui, element);
    }
    setProps(props) {
      this.placement = props.placement ?? this.placement;
      const oldProps = this.props;
      const el = this.element;
      if (el) {
        if (oldProps.className !== props.className) {
          if (oldProps.className)
            el.classList.remove(oldProps.className);
          if (props.className)
            el.classList.add(props.className);
        }
        if (!(0, import_core2._deepEqual)(oldProps.style, props.style, 1)) {
          (0, import_core2._removeStyles)(el, oldProps.style);
          (0, import_core2._applyStyles)(el, props.style);
        }
      }
      Object.assign(this.props, props);
      this.update();
    }
    getContainer() {
      return this.props.container || this.deck?.getCanvas()?.parentElement;
    }
    onFullscreenChange() {
      const prevFullscreen = this.fullscreen;
      const fullscreen = document.fullscreenElement === this.getContainer();
      if (prevFullscreen !== fullscreen) {
        this.fullscreen = !this.fullscreen;
      }
      this.update();
    }
    async handleClick() {
      if (this.fullscreen) {
        await this.exitFullscreen();
      } else {
        await this.requestFullscreen();
      }
      this.update();
    }
    async requestFullscreen() {
      const container = this.getContainer();
      if (container?.requestFullscreen) {
        await container.requestFullscreen({ navigationUI: "hide" });
      } else {
        this.togglePseudoFullscreen();
      }
    }
    async exitFullscreen() {
      if (document.exitFullscreen) {
        await document.exitFullscreen();
      } else {
        this.togglePseudoFullscreen();
      }
    }
    togglePseudoFullscreen() {
      this.getContainer()?.classList.toggle("deck-pseudo-fullscreen");
    }
  };

  // src/compass-widget.tsx
  var import_core3 = __toESM(require_core(), 1);
  var CompassWidget = class {
    constructor(props) {
      this.id = "compass";
      this.placement = "top-left";
      this.viewId = null;
      this.viewports = {};
      this.id = props.id ?? this.id;
      this.viewId = props.viewId ?? this.viewId;
      this.placement = props.placement ?? this.placement;
      this.props = {
        ...props,
        transitionDuration: props.transitionDuration ?? 200,
        label: props.label ?? "Reset Compass",
        style: props.style ?? {}
      };
    }
    setProps(props) {
      this.placement = props.placement ?? this.placement;
      this.viewId = props.viewId ?? this.viewId;
      const oldProps = this.props;
      const el = this.element;
      if (el) {
        if (oldProps.className !== props.className) {
          if (oldProps.className)
            el.classList.remove(oldProps.className);
          if (props.className)
            el.classList.add(props.className);
        }
        if (!(0, import_core3._deepEqual)(oldProps.style, props.style, 1)) {
          (0, import_core3._removeStyles)(el, oldProps.style);
          (0, import_core3._applyStyles)(el, props.style);
        }
      }
      Object.assign(this.props, props);
      this.update();
    }
    onViewportChange(viewport) {
      if (!viewport.equals(this.viewports[viewport.id])) {
        this.viewports[viewport.id] = viewport;
        this.update();
      }
    }
    onAdd({ deck }) {
      const { style, className } = this.props;
      const element = document.createElement("div");
      element.classList.add("deck-widget", "deck-widget-compass");
      if (className)
        element.classList.add(className);
      (0, import_core3._applyStyles)(element, style);
      this.deck = deck;
      this.element = element;
      this.update();
      return element;
    }
    getRotation(viewport) {
      if (viewport instanceof import_core3.WebMercatorViewport) {
        return [-viewport.bearing, viewport.pitch];
      } else if (viewport instanceof import_core3._GlobeViewport) {
        return [0, Math.max(-80, Math.min(80, viewport.latitude))];
      }
      return [0, 0];
    }
    update() {
      const viewId = this.viewId || Object.values(this.viewports)[0]?.id || "default-view";
      const viewport = this.viewports[viewId];
      const [rz, rx] = this.getRotation(viewport);
      const element = this.element;
      if (!element) {
        return;
      }
      const ui = /* @__PURE__ */ u2("div", { className: "deck-widget-button", style: { perspective: 100 }, children: /* @__PURE__ */ u2(
        "button",
        {
          type: "button",
          onClick: () => {
            for (const viewport2 of Object.values(this.viewports)) {
              this.handleCompassReset(viewport2);
            }
          },
          title: this.props.label,
          style: { transform: `rotateX(${rx}deg)` },
          children: /* @__PURE__ */ u2("svg", { fill: "none", width: "100%", height: "100%", viewBox: "0 0 26 26", children: /* @__PURE__ */ u2("g", { transform: `rotate(${rz},13,13)`, children: [
            /* @__PURE__ */ u2(
              "path",
              {
                d: "M10 13.0001L12.9999 5L15.9997 13.0001H10Z",
                fill: "var(--icon-compass-north-color, #F05C44)"
              }
            ),
            /* @__PURE__ */ u2(
              "path",
              {
                d: "M16.0002 12.9999L13.0004 21L10.0005 12.9999H16.0002Z",
                fill: "var(--icon-compass-south-color, #C2C2CC)"
              }
            )
          ] }) })
        }
      ) });
      B(ui, element);
    }
    onRemove() {
      this.deck = void 0;
      this.element = void 0;
    }
    handleCompassReset(viewport) {
      const viewId = this.viewId || viewport.id || "default-view";
      if (viewport instanceof import_core3.WebMercatorViewport) {
        const nextViewState = {
          ...viewport,
          bearing: 0,
          ...this.getRotation(viewport)[0] === 0 ? { pitch: 0 } : {},
          transitionDuration: this.props.transitionDuration,
          transitionInterpolator: new import_core3.FlyToInterpolator()
        };
        this.deck._onViewStateChange({ viewId, viewState: nextViewState, interactionState: {} });
      }
    }
  };

  // src/zoom-widget.tsx
  var import_core4 = __toESM(require_core(), 1);
  var _ZoomWidget = class {
    constructor(props = {}) {
      this.id = "zoom";
      this.placement = "top-left";
      this.viewId = null;
      this.viewports = {};
      this.id = props.id ?? this.id;
      this.viewId = props.viewId ?? this.viewId;
      this.placement = props.placement ?? this.placement;
      this.props = {
        ..._ZoomWidget.defaultProps,
        ...props
      };
    }
    onAdd({ deck }) {
      const { style, className } = this.props;
      const element = document.createElement("div");
      element.classList.add("deck-widget", "deck-widget-zoom");
      if (className)
        element.classList.add(className);
      (0, import_core4._applyStyles)(element, style);
      this.deck = deck;
      this.element = element;
      this.update();
      return element;
    }
    onRemove() {
      this.deck = void 0;
      this.element = void 0;
    }
    setProps(props) {
      this.placement = props.placement ?? this.placement;
      this.viewId = props.viewId ?? this.viewId;
      const oldProps = this.props;
      const el = this.element;
      if (el) {
        if (oldProps.className !== props.className) {
          if (oldProps.className)
            el.classList.remove(oldProps.className);
          if (props.className)
            el.classList.add(props.className);
        }
        if (!(0, import_core4._deepEqual)(oldProps.style, props.style, 1)) {
          (0, import_core4._removeStyles)(el, oldProps.style);
          (0, import_core4._applyStyles)(el, props.style);
        }
      }
      Object.assign(this.props, props);
      this.update();
    }
    onViewportChange(viewport) {
      this.viewports[viewport.id] = viewport;
    }
    handleZoom(viewport, nextZoom) {
      const viewId = this.viewId || viewport?.id || "default-view";
      const nextViewState = {
        ...viewport,
        zoom: nextZoom
      };
      if (this.props.transitionDuration > 0) {
        nextViewState.transitionDuration = this.props.transitionDuration;
        nextViewState.transitionInterpolator = "latitude" in nextViewState ? new import_core4.FlyToInterpolator() : new import_core4.LinearInterpolator();
      }
      this.setViewState(viewId, nextViewState);
    }
    handleZoomIn() {
      for (const viewport of Object.values(this.viewports)) {
        this.handleZoom(viewport, viewport.zoom + 1);
      }
    }
    handleZoomOut() {
      for (const viewport of Object.values(this.viewports)) {
        this.handleZoom(viewport, viewport.zoom - 1);
      }
    }
    /**
     * @todo - move to deck or widget manager
     */
    setViewState(viewId, viewState) {
      this.deck._onViewStateChange({ viewId, viewState, interactionState: {} });
    }
    update() {
      const element = this.element;
      if (!element) {
        return;
      }
      const ui = /* @__PURE__ */ u2(ButtonGroup, { orientation: this.props.orientation, children: [
        /* @__PURE__ */ u2(
          GroupedIconButton,
          {
            onClick: () => this.handleZoomIn(),
            label: this.props.zoomInLabel,
            className: "deck-widget-zoom-in"
          }
        ),
        /* @__PURE__ */ u2(
          GroupedIconButton,
          {
            onClick: () => this.handleZoomOut(),
            label: this.props.zoomOutLabel,
            className: "deck-widget-zoom-out"
          }
        )
      ] });
      B(ui, element);
    }
  };
  var ZoomWidget = _ZoomWidget;
  ZoomWidget.defaultProps = {
    id: "zoom-widget",
    style: {},
    placement: "top-left",
    className: void 0,
    orientation: "vertical",
    transitionDuration: 200,
    zoomInLabel: "Zoom In",
    zoomOutLabel: "Zoom Out",
    viewId: void 0
  };

  // src/widget-impl.ts
  var import_core5 = __toESM(require_core(), 1);
  var WidgetImpl = class {
    constructor(props) {
      this.id = props.id || "widget";
      this.props = props;
    }
    onAdd({ deck }) {
      this.deck = deck;
      const { style, className } = this.props;
      const el = this._createRootElement({
        widgetClassName: this.className,
        className,
        style
      });
      this.element = el;
      this.onRenderHTML();
      return this.element;
    }
    onRemove() {
      this.deck = void 0;
      this.element = void 0;
    }
    setProps(props) {
      const oldProps = this.props;
      const el = this.element;
      if (el) {
        if (oldProps.className !== props.className) {
          if (oldProps.className)
            el.classList.remove(oldProps.className);
          if (props.className)
            el.classList.add(props.className);
        }
        if (!(0, import_core5._deepEqual)(oldProps.style, props.style, 1)) {
          (0, import_core5._removeStyles)(el, oldProps.style);
          (0, import_core5._applyStyles)(el, props.style);
        }
      }
      Object.assign(this.props, props);
      this.onRenderHTML();
    }
    _createRootElement(props) {
      const { widgetClassName, className, style } = props;
      const element = document.createElement("div");
      ["deck-widget", widgetClassName, className].filter((cls) => typeof cls === "string" && cls.length > 0).forEach((className2) => element.classList.add(className2));
      (0, import_core5._applyStyles)(element, style);
      return element;
    }
  };
  WidgetImpl.defaultProps = {
    id: "widget",
    style: {},
    className: ""
  };

  // src/screenshot-widget.tsx
  var _ScreenshotWidget = class extends WidgetImpl {
    constructor(props = {}) {
      super({ ..._ScreenshotWidget.defaultProps, ...props });
      this.className = "deck-widget-screenshot";
      this.placement = "top-left";
      this.placement = props.placement ?? this.placement;
    }
    setProps(props) {
      this.placement = props.placement ?? this.placement;
      super.setProps(props);
    }
    onRenderHTML() {
      const element = this.element;
      if (!element)
        return;
      B(
        /* @__PURE__ */ u2(
          IconButton,
          {
            className: "deck-widget-camera",
            label: this.props.label,
            onClick: this.handleClick.bind(this)
          }
        ),
        element
      );
    }
    handleClick() {
      if (this.props.onCapture) {
        this.props.onCapture(this);
        return;
      }
      const dataURL = this.captureScreenToDataURL(this.props.imageFormat);
      if (dataURL) {
        this.downloadDataURL(dataURL, this.props.filename);
      }
    }
    /** @note only captures canvas contents, not HTML DOM or CSS styles */
    captureScreenToDataURL(imageFormat) {
      const canvas = this.deck?.getCanvas();
      return canvas?.toDataURL(imageFormat);
    }
    /** Download a data URL */
    downloadDataURL(dataURL, filename) {
      const link = document.createElement("a");
      link.href = dataURL;
      link.download = filename;
      link.click();
    }
  };
  var ScreenshotWidget = _ScreenshotWidget;
  ScreenshotWidget.defaultProps = {
    ...WidgetImpl.defaultProps,
    id: "screenshot",
    placement: "top-left",
    label: "Screenshot",
    filename: "screenshot.png",
    imageFormat: "image/png",
    onCapture: void 0
  };

  // src/reset-view-widget.tsx
  var _ResetViewWidget = class extends WidgetImpl {
    constructor(props = {}) {
      super({ ..._ResetViewWidget.defaultProps, ...props });
      this.className = "deck-widget-reset-view";
      this.placement = "top-left";
      this.placement = props.placement ?? this.placement;
    }
    setProps(props) {
      this.placement = props.placement ?? this.placement;
      super.setProps(props);
    }
    onRenderHTML() {
      const element = this.element;
      if (!element)
        return;
      B(
        /* @__PURE__ */ u2(
          IconButton,
          {
            className: "deck-widget-reset-focus",
            label: this.props.label,
            onClick: this.handleClick.bind(this)
          }
        ),
        element
      );
    }
    handleClick() {
      const initialViewState = this.props.initialViewState || this.deck?.props.initialViewState;
      this.setViewState(initialViewState);
    }
    setViewState(viewState) {
      const viewId = this.props.viewId || viewState?.id || "default-view";
      const nextViewState = {
        ...viewState
        // only works for geospatial?
        // transitionDuration: this.props.transitionDuration,
        // transitionInterpolator: new FlyToInterpolator()
      };
      this.deck._onViewStateChange({ viewId, viewState: nextViewState, interactionState: {} });
    }
  };
  var ResetViewWidget = _ResetViewWidget;
  ResetViewWidget.defaultProps = {
    ...WidgetImpl.defaultProps,
    id: "reset-view",
    placement: "top-left",
    label: "Reset View",
    initialViewState: void 0,
    viewId: void 0
  };

  // src/themes.ts
  var LightTheme = {
    "--button-background": "#fff",
    "--button-stroke": "rgba(255, 255, 255, 0.3)",
    "--button-inner-stroke": "unset",
    "--button-shadow": "0px 0px 8px 0px rgba(0, 0, 0, 0.25)",
    "--button-backdrop-filter": "unset",
    "--button-icon-idle": "rgba(97, 97, 102, 1)",
    "--button-icon-hover": "rgba(24, 24, 26, 1)",
    "--icon-compass-north-color": "#F05C44",
    "--icon-compass-south-color": "#C2C2CC"
  };
  var DarkTheme = {
    "--button-background": "rgba(18, 18, 20, 1)",
    "--button-stroke": "rgba(18, 18, 20, 0.30)",
    "--button-inner-stroke": "unset",
    "--button-shadow": "0px 0px 8px 0px rgba(0, 0, 0, 0.25)",
    "--button-backdrop-filter": "unset",
    "--button-icon-idle": "rgba(158, 157, 168, 1)",
    "--button-icon-hover": "rgba(215, 214, 229, 1)",
    "--icon-compass-north-color": "#F05C44",
    "--icon-compass-south-color": "#C2C2CC"
  };
  var LightGlassTheme = {
    "--button-background": "rgba(255, 255, 255, 0.6)",
    "--button-stroke": "rgba(255, 255, 255, 0.3)",
    "--button-inner-stroke": "1px solid rgba(255, 255, 255, 0.6)",
    "--button-shadow": "0px 0px 8px 0px rgba(0, 0, 0, 0.25), 0px 0px 8px 0px rgba(0, 0, 0, 0.1) inset",
    "--button-backdrop-filter": "blur(4px)",
    "--button-icon-idle": "rgba(97, 97, 102, 1)",
    "--button-icon-hover": "rgba(24, 24, 26, 1)",
    "--icon-compass-north-color": "#F05C44",
    "--icon-compass-south-color": "#C2C2CC"
  };
  var DarkGlassTheme = {
    "--button-background": "rgba(18, 18, 20, 0.75)",
    "--button-stroke": "rgba(18, 18, 20, 0.30)",
    "--button-inner-stroke": "1px solid rgba(18, 18, 20, 0.75)",
    "--button-shadow": "0px 0px 8px 0px rgba(0, 0, 0, 0.25), 0px 0px 8px 0px rgba(0, 0, 0, 0.1) inset",
    "--button-backdrop-filter": "blur(4px)",
    "--button-icon-idle": "rgba(158, 157, 168, 1)",
    "--button-icon-hover": "rgba(215, 214, 229, 1)",
    "--icon-compass-north-color": "#F05C44",
    "--icon-compass-south-color": "#C2C2CC"
  };
  return __toCommonJS(bundle_exports);
})();
      return __exports__;
      });
