// @deck.gl/widgets/dist/reset-view-widget.js
import { jsx as _jsx } from "preact/jsx-runtime";
import { render } from 'preact';
import { WidgetImpl } from "./widget-impl.js";
import { IconButton } from "./components.js";
/**
 * A button widget that resets the view state of deck to an initial state.
 */
export class ResetViewWidget extends WidgetImpl {
    constructor(props = {}) {
        super({ ...ResetViewWidget.defaultProps, ...props });
        this.className = 'deck-widget-reset-view';
        this.placement = 'top-left';
        this.placement = props.placement ?? this.placement;
    }
    setProps(props) {
        this.placement = props.placement ?? this.placement;
        super.setProps(props);
    }
    onRenderHTML() {
        const element = this.element;
        if (!element)
            return;
        render(_jsx(IconButton, { className: "deck-widget-reset-focus", label: this.props.label, onClick: this.handleClick.bind(this) }), element);
    }
    handleClick() {
        const initialViewState = this.props.initialViewState || this.deck?.props.initialViewState;
        this.setViewState(initialViewState);
    }
    setViewState(viewState) {
        const viewId = this.props.viewId || viewState?.id || 'default-view';
        const nextViewState = {
            ...viewState
            // only works for geospatial?
            // transitionDuration: this.props.transitionDuration,
            // transitionInterpolator: new FlyToInterpolator()
        };
        // @ts-ignore Using private method temporary until there's a public one
        this.deck._onViewStateChange({ viewId, viewState: nextViewState, interactionState: {} });
    }
}
ResetViewWidget.defaultProps = {
    ...WidgetImpl.defaultProps,
    id: 'reset-view',
    placement: 'top-left',
    label: 'Reset View',
    initialViewState: undefined,
    viewId: undefined
};
