// @deck.gl/widgets/dist/screenshot-widget.js
import { jsx as _jsx } from "preact/jsx-runtime";
// deck.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
/* global document */
import { WidgetImpl } from "./widget-impl.js";
import { render } from 'preact';
import { IconButton } from "./components.js";
/**
 * A button widget that captures a screenshot of the current canvas and downloads it as a (png) file.
 * @note only captures canvas contents, not HTML DOM or CSS styles
 */
export class ScreenshotWidget extends WidgetImpl {
    constructor(props = {}) {
        super({ ...ScreenshotWidget.defaultProps, ...props });
        this.className = 'deck-widget-screenshot';
        this.placement = 'top-left';
        this.placement = props.placement ?? this.placement;
    }
    setProps(props) {
        this.placement = props.placement ?? this.placement;
        super.setProps(props);
    }
    onRenderHTML() {
        const element = this.element;
        if (!element)
            return;
        render(_jsx(IconButton, { className: "deck-widget-camera", label: this.props.label, onClick: this.handleClick.bind(this) }), element);
    }
    handleClick() {
        // Allow user to override the capture logic
        if (this.props.onCapture) {
            this.props.onCapture(this);
            return;
        }
        const dataURL = this.captureScreenToDataURL(this.props.imageFormat);
        if (dataURL) {
            this.downloadDataURL(dataURL, this.props.filename);
        }
    }
    /** @note only captures canvas contents, not HTML DOM or CSS styles */
    captureScreenToDataURL(imageFormat) {
        const canvas = this.deck?.getCanvas();
        return canvas?.toDataURL(imageFormat);
    }
    /** Download a data URL */
    downloadDataURL(dataURL, filename) {
        const link = document.createElement('a');
        link.href = dataURL;
        link.download = filename;
        link.click();
    }
}
ScreenshotWidget.defaultProps = {
    ...WidgetImpl.defaultProps,
    id: 'screenshot',
    placement: 'top-left',
    label: 'Screenshot',
    filename: 'screenshot.png',
    imageFormat: 'image/png',
    onCapture: undefined
};
